R/test.R

```diff
@@ -0,0 +1,107 @@
+
+#' Check to see if the given text is a shinytest
+#' Scans for the magic string of `app <- ShinyDriver$new(` as an indicator that this is a shinytest.
+#' Brought in from shinytest to avoid having to export this function.
+#' @noRd
+isShinyTest <- function(text){
+  lines <- grepl("app\\s*<-\\s*ShinyDriver\\$new\\(", text, perl=TRUE)
+  any(lines)
+}
+
+#' Runs the tests associated with this Shiny app
+#'
+#' Sources the `.R` files in the top-level of `tests/` much like `R CMD check`.
+#' These files are typically simple runners for tests nested in other
+#' directories under `tests/`.
+#'
+#' @param appDir The base directory for the application.
+#' @param filter If not `NULL`, only tests with file names matching this regular
+#'   expression will be executed passed to `grepl`. Matching is performed on
+#'   the file name including the extension.
+#'
+#' @details Historically, [shinytest](https://rstudio.github.io/shinytest/)
+#'   recommended placing tests at the top-level of the `tests/` directory. In
+#'   order to support that model, `testApp` first checks to see if the `.R`
+#'   files in the `tests/` directory are all shinytests; if so, just calls out
+#'   to [shinytest::testApp()].
+#' @export
+runTests <- function(appDir=".", filter=NULL){
+  require(shiny)
+
+  testsDir <- file.path(appDir, "tests")
+  if (!dirExists(testsDir)){
+    stop("No tests directory found: ", testsDir)
+  }
+  runners <- list.files(testsDir, pattern="\\.r$", ignore.case = TRUE)
+
+  if (length(runners) == 0){
+    message("No test runners found in ", testsDir)
+    return(structure(list(result=NA, files=list()), class="shinytestrun"))
+  }
+
+  if (!is.null(filter)){
+    runners <- runners[grepl(filter, runners)]
+  }
+  if (length(runners) == 0){
+    stop("No test runners matched the given filter: '", filter, "'")
+  }
+
+  # Inspect each runner to see if it appears to be a shinytest
+  isST <- vapply(runners, function(r){
+    text <- readLines(file.path(testsDir, r), warn = FALSE)
+    isShinyTest(text)
+  }, logical(1))
+
+  if (all(isST)){
+    # just call out to shinytest
+    # We don't need to message/warn here since shinytest already does it.
+    if (!requireNamespace("shinytest", quietly=TRUE) ){
+      stop("It appears that the .R files in ", testsDir,
+           " are all shinytests, but shinytest is not installed.")
+    }
+
+    if (getOption("shiny.autoload.r", FALSE)) {
+      warning("You've enabled `shiny.autoload.r` but this is not supported yet with shinytest.")
```

Comment:
This message occurs if the user has explicitly set `options(shiny.autoload.r=T)` in this R session, but is using Shiny 1.4.0, where it defaults to `F`, right? I think that for the upcoming version of Shiny, where it defaults to `T`, then this message shouldn't come up.

If that's right, then this message should be tweaked a bit to explain what's going on. We could also make shinytest pass along this R session's value for `shiny.autoload.r` by default, but it would be simpler to just get users to upgrade their Shiny version.



This message occurs if the user has explicitly set `options(shiny.autoload.r=T)` in this R session, but is using Shiny 1.4.0, where it defaults to `F`, right? I think that for the upcoming version of Shiny, where it defaults to `T`, then this message shouldn't come up.

If that's right, then this message should be tweaked a bit to explain what's going on. We could also make shinytest pass along this R session's value for `shiny.autoload.r` by default, but it would be simpler to just get users to upgrade their Shiny version.